plugin/code_actions.py

```diff
@@ -333,3 +339,43 @@ def run_async() -> None:
     def _handle_response_async(self, session_name: str, response: Any) -> None:
         if isinstance(response, Error):
             sublime.error_message("{}: {}".format(session_name, str(response)))
+
+
+class LspRefactorCommand(LspTextCommand):
+
+    capability = 'codeActionProvider'
+
+    def is_visible(self, id: int, event: Optional[dict] = None, point: Optional[int] = None) -> bool:
+        return len(actions_manager.refactor_actions_cache) > id
+
+    def description(self, id: int, event: Optional[dict] = None, point: Optional[int] = None) -> Optional[str]:
+        if self.is_visible(id):
+            return actions_manager.refactor_actions_cache[id][1]['title']
+
+    def run(self, edit: sublime.Edit, id: int, event: Optional[dict] = None, point: Optional[int] = None) -> None:
+        sublime.set_timeout_async(partial(self.run_async, id))
+
+    def _is_cache_valid(self) -> bool:
+        v = self.view
+        if actions_manager._response_cache is None:
+            return False
+        region = first_selection_region(v)
+        if region is None:
+            return False
+        listener = windows.listener_for_view(v)
+        if not listener:
+            return False
+        _, covering = listener.diagnostics_intersecting_async(region)
+        return actions_manager._response_cache[0] == "{}#{}:{}".format(v.buffer_id(), v.change_count(), covering)
+
+    def run_async(self, id: int) -> None:
+        if self._is_cache_valid():
+            config_name, action = actions_manager.refactor_actions_cache[id]
+            session = self.session_by_name(config_name)
+            if session:
+                session.run_code_action_async(action, progress=True) \
+                    .then(lambda response: self._handle_response_async(config_name, response))
+
+    def _handle_response_async(self, session_name: str, response: Any) -> None:
```

Comment:
A method order like:
- run
- run_async
- _handle_response_async
- _is_cache_valid

would be better (or with two last reversed)  since that would better match the calling order.